cpp/src/parquet/column_reader_test.cc

```diff
@@ -572,5 +576,549 @@ TEST_F(TestPrimitiveReader, TestNonDictionaryEncodedPagesWithExposeEncoding) {
+  std::vector<int16_t> rep_levels(2048, 0);
+
+  std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+      descr_.get(), values, /*num_values=*/static_cast<int>(values.size()),
+      Encoding::PLAIN,
+      /*indices=*/{},
+      /*indices_size=*/0, def_levels, level_info_.def_level, rep_levels,
+      level_info_.rep_level);
+  pages.push_back(std::move(page));
+  auto pager = std::make_unique<MockPageReader>(pages);
+  record_reader_->SetPageReader(std::move(pager));
+  {
+    // Read 1000 records. We will read 1024 levels because that is the minimum
+    // number of levels to read.
+    int64_t records_read = record_reader_->ReadRecords(/*num_records=*/1000);
+    ASSERT_EQ(records_read, 1000);
+    CheckState(/*values_written=*/1000, /*null_count=*/0, /*levels_written=*/1024,
+               /*levels_position=*/1000);
+    std::vector<int32_t> expected_values(1000, 10);
+    std::vector<int16_t> expected_def_levels(1000, 1);
+    std::vector<int16_t> expected_rep_levels(1000, 0);
+    CheckReadValues(expected_values, expected_def_levels, expected_rep_levels);
+    // Reset removes the already consumed values and levels.
+    record_reader_->Reset();
+  }
+
+  {  // Skip 12 records. Since we already have 24 in the buffer, we should not be
+    // reading any more levels into the buffer, we will just consume 12 of it.
+    int64_t records_skipped = record_reader_->SkipRecords(/*num_records=*/12);
+    ASSERT_EQ(records_skipped, 12);
+    CheckState(/*values_written=*/0, /*null_count=*/0, /*levels_written=*/12,
+               /*levels_position=*/0);
+    // Everthing is empty because we reset the reader before this skip.
+    CheckReadValues(/*expected_values=*/{}, /*expected_def_levels=*/{},
+                    /*expected_rep_levels=*/{});
+  }
+}
+
+// Test reading when one record spans multiple pages for a repeated field.
+TEST_F(RecordReaderTest, ReadPartialRecord) {
+  Init(/*max_def_level=*/1, /*max_rep_level=*/1, Repetition::REPEATED);
+
+  std::vector<std::shared_ptr<Page>> pages;
+
+  // Page 1: {[10], [20, 20, 20 ... } continues to next page.
+  {
+    std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+        descr_.get(), /*values=*/{10, 20, 20, 20}, /*num_values=*/4, Encoding::PLAIN,
+        /*indices=*/{},
+        /*indices_size=*/0, /*def_levels=*/{1, 1, 1, 1}, level_info_.def_level,
+        /*rep_levels=*/{0, 0, 1, 1}, level_info_.rep_level);
+    pages.push_back(std::move(page));
+  }
+
+  // Page 2: {... 20, 20, ...} continues from previous page and to next page.
+  {
+    std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+        descr_.get(), /*values=*/{20, 20}, /*num_values=*/2, Encoding::PLAIN,
+        /*indices=*/{},
+        /*indices_size=*/0, /*def_levels=*/{1, 1}, level_info_.def_level,
+        /*rep_levels=*/{1, 1}, level_info_.rep_level);
+    pages.push_back(std::move(page));
+  }
+
+  // Page 3: { ... 20], [30]} continues from previous page.
```

Comment:
This is already a neat example, but can we add one or two nulls in the mix? :-)

Reply:
I already tested for skipping nulls in SkipRepeated. I am not sure if adding null records to this test will increase coverage. Since a null record will not span multiple pages.